### ISSUE TITLE
Unknown encoder 'libx264'

### ISSUE THREAD

**Issue by mike1011** [shunjikonishi/heroku-buildpack-ffmpeg#3]
this is my .buildpacks

https://github.com/shunjikonishi/heroku-buildpack-ffmpeg
https://github.com/heroku/heroku-buildpack-ruby.git

but it says that the libx264 is missing.so how can i explicitly install and include it.Below is my log from papertrail..

 Apr 21 13:06:01 video-convertor app/web.1: cache: [GET /assets/delete.png] miss
Apr 21 13:08:01 video-convertor app/web.1: Started POST "/videos" for 180.188.254.229 at 2014-04-21 20:08:00 +0000
Apr 21 13:08:01 video-convertor app/web.1: Processing by VideosController#create as HTML
Apr 21 13:08:01 video-convertor app/web.1: "create==========="
Apr 21 13:08:01 video-convertor app/web.1: {"utf8"=>"✓", "authenticity_token"=>"+IbL4Hc9cpnip6kImnSx13UikqAkXzXDpGw98gWKxe0=", "video"=>{"avatar"=>#<ActionDispatch::Http::UploadedFile:0x000000010963b0 @original_filename="Wildlife.wmv", @content_type="video/x-ms-wmv", @headers="Content-Disposition: form-data; name=\"video[avatar]\"; filename=\"Wildlife.wmv\"\r\nContent-Type: video/x-ms-wmv\r\n", @tempfile=#<File:/tmp/RackMultipart20140421-2-m92l6l>>, "convert"=>"flv"}, "action"=>"create", "controller"=>"videos"}
Apr 21 13:08:01 video-convertor app/web.1: [paperclip] [ffmpeg] ffprobe "/tmp/Wildlife20140421-2-1k8rhl7.wmv" 2>&1
Apr 21 13:08:01 video-convertor app/web.1:   Parameters: {"utf8"=>"✓", "authenticity_token"=>"+IbL4Hc9cpnip6kImnSx13UikqAkXzXDpGw98gWKxe0=", "video"=>{"avatar"=>#<ActionDispatch::Http::UploadedFile:0x000000010963b0 @original_filename="Wildlife.wmv", @content_type="video/x-ms-wmv", @headers="Content-Disposition: form-data; name=\"video[avatar]\"; filename=\"Wildlife.wmv\"\r\nContent-Type: video/x-ms-wmv\r\n", @tempfile=#<File:/tmp/RackMultipart20140421-2-m92l6l>>, "convert"=>"flv"}}
Apr 21 13:08:02 video-convertor app/web.1:   libavutil 52. 34.100 / 52. 34.100
Apr 21 13:08:02 video-convertor app/web.1:   libavcodec 55. 13.100 / 55. 13.100
Apr 21 13:08:02 video-convertor app/web.1:   libavformat 55. 8.102 / 55. 8.102
Apr 21 13:08:02 video-convertor app/web.1:   libavdevice 55. 2.100 / 55. 2.100
Apr 21 13:08:02 video-convertor app/web.1:   libavfilter 3. 74.101 / 3. 74.101
Apr 21 13:08:02 video-convertor app/web.1:   libswscale 2. 3.100 / 2. 3.100
Apr 21 13:08:02 video-convertor app/web.1:   libswresample 0. 17.102 / 0. 17.102
Apr 21 13:08:02 video-convertor app/web.1: [asf @ 0x1545850] Stream #0: not enough frames to estimate rate; consider increasing probesize
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] An error was received while processing: #<Paperclip::Error: error while processing video for Wildlife20140421-2-1k8rhl7: Command 'ffmpeg -i '/tmp/Wildlife20140421-2-1k8rhl7.wmv' -vcodec libvpx -b 250k -bt 50k -acodec libvorbis -ab 56k -ac 2 -s 520x292 -f webm -y '/tmp/Wildlife20140421-2-1k8rhl720140421-2-p4g8fn.webm'' returned 1. Expected 0
Apr 21 13:08:02 video-convertor app/web.1: Here is the command output:
Apr 21 13:08:02 video-convertor heroku/router: at=info method=POST path=/videos host=video-convertor.herokuapp.com request_id=a84551e4-c359-421d-af11-43bcef88ba2c fwd="180.188.254.229" dyno=web.1 connect=2ms service=111112ms status=200 bytes=7821
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Command Success
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Making...
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Building Destination File: 'Wildlife20140421-2-1k8rhl7' + 'webm'
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Destination File Built
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Adding Geometry
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Extracting Target Dimensions
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Target Size is Available
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Keeping Aspect Ratio
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Resize
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Convert Options: 520x292
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Adding Format
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Adding Source
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Building Parameters
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] -i :source -vcodec libvpx -b 250k -bt 50k -acodec libvorbis -ab 56k -ac 2 -s 520x292 -f webm -y :dest
Apr 21 13:08:02 video-convertor app/web.1: Command :: ffmpeg -i '/tmp/Wildlife20140421-2-1k8rhl7.wmv' -vcodec libvpx -b 250k -bt 50k -acodec libvorbis -ab 56k -ac 2 -s 520x292 -f webm -y '/tmp/Wildlife20140421-2-1k8rhl720140421-2-p4g8fn.webm'
Apr 21 13:08:02 video-convertor app/web.1: Guessed Channel Layout for Input Stream #0.0 : stereo
Apr 21 13:08:02 video-convertor app/web.1: Input #0, asf, from '/tmp/Wildlife20140421-2-1k8rhl7.wmv':
Apr 21 13:08:02 video-convertor app/web.1:   Metadata:
Apr 21 13:08:02 video-convertor app/web.1:     SfOriginalFPS : 299700
Apr 21 13:08:02 video-convertor app/web.1:     WMFSDKVersion : 11.0.6001.7000
Apr 21 13:08:02 video-convertor app/web.1:     WMFSDKNeeded : 0.0.0.0000
Apr 21 13:08:02 video-convertor app/web.1:     comment : Footage: Small World Productions, Inc; Tourism New Zealand | Producer: Gary F. Spradling | Music: Steve Ball
Apr 21 13:08:02 video-convertor app/web.1:     title : Wildlife in HD
Apr 21 13:08:02 video-convertor app/web.1:     copyright : © 2008 Microsoft Corporation
Apr 21 13:08:02 video-convertor app/web.1:     IsVBR : 0
Apr 21 13:08:02 video-convertor app/web.1:     DeviceConformanceTemplate: AP@L3
Apr 21 13:08:02 video-convertor app/web.1:   Duration: 00:00:30.09, start: 0.000000, bitrate: 6977 kb/s
Apr 21 13:08:02 video-convertor app/web.1:     Stream #0:0(eng): Audio: wmav2 (a[1][0][0] / 0x0161), 44100 Hz, stereo, fltp, 192 kb/s
Apr 21 13:08:02 video-convertor app/web.1: >
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] ffprobe "/tmp/Wildlife20140421-2-1k8rhl7.wmv" 2>&1
Apr 21 13:08:02 video-convertor app/web.1: ffmpeg version git-2013-06-02-5711e4f Copyright (c) 2000-2013 the FFmpeg developers
Apr 21 13:08:02 video-convertor app/web.1:   built on Jun 2 2013 07:38:40 with gcc 4.4.3 (Ubuntu 4.4.3-4ubuntu5.1)
Apr 21 13:08:02 video-convertor app/web.1:   configuration: --enable-shared --disable-asm --prefix=/app/vendor/ffmpeg
Apr 21 13:08:02 video-convertor app/web.1:   libavutil 52. 34.100 / 52. 34.100
Apr 21 13:08:02 video-convertor app/web.1:   libavcodec 55. 13.100 / 55. 13.100
Apr 21 13:08:02 video-convertor app/web.1:   libavformat 55. 8.102 / 55. 8.102
Apr 21 13:08:02 video-convertor app/web.1:   libavdevice 55. 2.100 / 55. 2.100
Apr 21 13:08:02 video-convertor app/web.1:   libavfilter 3. 74.101 / 3. 74.101
Apr 21 13:08:02 video-convertor app/web.1:   libswscale 2. 3.100 / 2. 3.100
Apr 21 13:08:02 video-convertor app/web.1:   libswresample 0. 17.102 / 0. 17.102
Apr 21 13:08:02 video-convertor app/web.1:     Stream #0:1(eng): Video: vc1 (Advanced) (WVC1 / 0x31435657), yuv420p, 1280x720, 5942 kb/s, 29.97 tbr, 1k tbn, 1k tbc
Apr 21 13:08:02 video-convertor app/web.1: Please use -b:a or -b:v, -b is ambiguous
Apr 21 13:08:02 video-convertor app/web.1: Unknown encoder 'libvpx'
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Command Success
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Making...
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Building Destination File: 'Wildlife20140421-2-1k8rhl7' + 'flv'
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Destination File Built
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Adding Geometry
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Extracting Target Dimensions
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Target Size is Available
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Keeping Aspect Ratio
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Resize
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Convert Options: 520x292
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Adding Format
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Adding Source
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] Building Parameters
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] [ffmpeg] -i :source -vcodec libx264 -b 1000000 -ar 22050 -acodec mp3 -ab 56k -ac 2 -s 520x292 -y :dest
Apr 21 13:08:02 video-convertor app/web.1: Command :: ffmpeg -i '/tmp/Wildlife20140421-2-1k8rhl7.wmv' -vcodec libx264 -b 1000000 -ar 22050 -acodec mp3 -ab 56k -ac 2 -s 520x292 -y '/tmp/Wildlife20140421-2-1k8rhl720140421-2-vo3uhl.flv'
Apr 21 13:08:02 video-convertor app/web.1: [asf @ 0x228f830] Stream #0: not enough frames to estimate rate; consider increasing probesize
Apr 21 13:08:02 video-convertor app/web.1: [paperclip] An error was received while processing: #<Paperclip::Error: error while processing video for Wildlife20140421-2-1k8rhl7: Command 'ffmpeg -i '/tmp/Wildlife20140421-2-1k8rhl7.wmv' -vcodec libx264 -b 1000000 -ar 22050 -acodec mp3 -ab 56k -ac 2 -s 520x292 -y '/tmp/Wildlife20140421-2-1k8rhl720140421-2-vo3uhl.flv'' returned 1. Expected 0
Apr 21 13:08:02 video-convertor app/web.1: Here is the command output:
Apr 21 13:08:02 video-convertor app/web.1: >
Apr 21 13:08:02 video-convertor app/web.1: [deprecated] I18n.enforce_available_locales will default to true in the future. If you really want to skip validation of your locale you can set I18n.enforce_available_locales = false to avoid this message.
Apr 21 13:08:02 video-convertor app/web.1: cache: [POST /videos] invalidate, pass
Apr 21 13:08:02 video-convertor app/web.1: Guessed Channel Layout for Input Stream #0.0 : stereo
Apr 21 13:08:02 video-convertor app/web.1: Input #0, asf, from '/tmp/Wildlife20140421-2-1k8rhl7.wmv':
Apr 21 13:08:02 video-convertor app/web.1:   Metadata:
Apr 21 13:08:02 video-convertor app/web.1:     SfOriginalFPS : 299700
Apr 21 13:08:02 video-convertor app/web.1:     WMFSDKVersion : 11.0.6001.7000
Apr 21 13:08:02 video-convertor app/web.1:     WMFSDKNeeded : 0.0.0.0000
Apr 21 13:08:02 video-convertor app/web.1:     comment : Footage: Small World Productions, Inc; Tourism New Zealand | Producer: Gary F. Spradling | Music: Steve Ball
Apr 21 13:08:02 video-convertor app/web.1:     title : Wildlife in HD
Apr 21 13:08:02 video-convertor app/web.1:     copyright : © 2008 Microsoft Corporation
Apr 21 13:08:02 video-convertor app/web.1:     IsVBR : 0
Apr 21 13:08:02 video-convertor app/web.1:     DeviceConformanceTemplate: AP@L3
Apr 21 13:08:02 video-convertor app/web.1:   Duration: 00:00:30.09, start: 0.000000, bitrate: 6977 kb/s
Apr 21 13:08:02 video-convertor app/web.1:     Stream #0:0(eng): Audio: wmav2 (a[1][0][0] / 0x0161), 44100 Hz, stereo, fltp, 192 kb/s
Apr 21 13:08:02 video-convertor app/web.1:     Stream #0:1(eng): Video: vc1 (Advanced) (WVC1 / 0x31435657), yuv420p, 1280x720, 5942 kb/s, 29.97 tbr, 1k tbn, 1k tbc
Apr 21 13:08:02 video-convertor app/web.1: Please use -b:a or -b:v, -b is ambiguous
Apr 21 13:08:02 video-convertor app/web.1: Unknown encoder 'libx264'
Apr 21 13:08:02 video-convertor app/web.1:   Rendered videos/_form.html.erb (10.8ms)
Apr 21 13:08:02 video-convertor app/web.1:   Rendered videos/new.html.erb within layouts/application (11.4ms)

**Comment by shunjikonishi** [comment]
Hi,  
Probably this is issue of ffmpeg binary.  
My ffmpeg binary was built at one year ago.  

Maybe libx264 support of ffmpeg is newer than it.

So, if you want to use new features of ffmpeg, you should fork this buildpack and use your own ffmpeg binary.

Regards.

**Comment by mike1011** [comment]
thanks for the reply @shunjikonishi ....can u give me a hint as where i should fork it to get the required compiled codec.

**Comment by snehasishroy** [comment]
@shunjikonishi I am also facing the same issue. I build my own ffmpeg binary and then I forked your repo and edited the download _url. Have a look : https://github.com/snehasishroy/heroku-buildpack-ffmpeg/blob/master/bin/compile

In the .buildpacks I mentioned my github repo along with heroku php repo

After running ffmpeg -version it said that ffmpeg command not found
Can you kindly tell me what more should I do ? Can you check my ffmpeg tar file ?

**Comment by shunjikonishi** [comment]
Hi, I tried your buildpack.
And found following message.

```
gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Exiting with failure status due to previous errors
```

Maybe your archive file is not gzip.

You should do either of followings.
- Replace archive file  to gzip format.
- Change unpack command

https://github.com/snehasishroy/heroku-buildpack-ffmpeg/blob/master/bin/compile#L17

Best.

**Comment by snehasishroy** [comment]
Hi,
Thanks for pointing that out. I updated the download_url. Now when I run ffmpeg -version, it states 

> error while loading shared libraries: libva.so.1: cannot open shared object file.No such file or directory.

I compiled my ffmpeg binaries on 64 bit Ubuntu 14.04 using the steps mentioned here. 

> https://trac.ffmpeg.org/wiki/CompilationGuide/Ubuntu

I think the problem is with the dependencies associated with the binary. Can you kindly point me out in the right direction.

**Comment by snehasishiiita** [comment]
Never mind :) Solved the problem. I compiled my ffmpeg on Ubuntu 10.04 64 bit and removed unnecessary dependencies and everything works fine. Thanks :+1:

**Comment by shunjikonishi** [comment]
All right!

**Comment by JetStarBlues** [comment]
Hi

@snehasishiiita
I had the same `libva.so.1` no such file error.
I tried to use the binary you [created](https://github.com/snehasishroy/heroku-buildpack-ffmpeg)

But I now get the following error...
`ffmpeg: error while loading shared libraries: libvorbisenc.so.2: cannot open shared object file: No such file or directory`

From your post, it seems the binary worked for you... any idea what's going on?

**EDIT:** 

Nevermind... I figured out how to properly compile from source. I was using `sudo apt-get` for things like `libtheora` and `libvorbis`.
